torch/csrc/jit/passes/frozen_concat_linear.cpp

```diff
@@ -0,0 +1,249 @@
+      Node* base_node = linear_layer_group[i];
+      if (checked_nodes.find(base_node) != checked_nodes.end()) {
+        continue;
+      }
+
+      std::vector<Node*> compatible_layers;
+      compatible_layers.push_back(base_node);
+      checked_nodes.insert(base_node);
+
+      auto base_weight =
+          constant_as<Tensor>(base_node->namedInput("weight")).value();
+      auto base_bias =
+          constant_as<Tensor>(base_node->namedInput("bias")).value();
+
+      // Now iterate over the rest of the users of the set to
+      // see if there is anything that we can coaleasce `base_node` with.
+      for (size_t j = i + 1; j < linear_layer_group.size(); j++) {
+        auto node = linear_layer_group[j];
+        if (checked_nodes.find(node) != checked_nodes.end()) {
+          continue;
+        }
+        auto weight = constant_as<Tensor>(node->namedInput("weight")).value();
+        auto bias = constant_as<Tensor>(node->namedInput("bias")).value();
+
+        // For now we will just keep it simple and require matching types
+        // Type promotion might cause performance to actually decrease.
+        if (base_weight.dtype() != weight.dtype() ||
+            base_weight.device() != weight.device() ||
+            base_bias.dtype() != bias.dtype() ||
+            base_bias.device() != bias.device()) {
+          continue;
+        }
+
+        if (!isNonZeroDimEqual(base_weight, weight) ||
+            !isNonZeroDimEqual(base_bias, bias)) {
+          continue;
+        }
+
+        bool can_move_before_all = true;
+        for (auto n : compatible_layers) {
+          can_move_before_all &=
+              getAliasDb()->couldMoveBeforeTopologically(node, n);
+        }
+        if (!can_move_before_all) {
+          continue;
+        }
+
+        // Found a node that is eligible for combination
+        compatible_layers.push_back(node);
+        checked_nodes.insert(node);
+      }
+      if (compatible_layers.size() == 1) {
+        continue; // No other layers to merge
+      }
+      mergeLinearLayers(compatible_layers);
+    }
+  }
+
+  void handleBlockAndSubblocks(Block* block) {
+    for (auto node : block->nodes()) {
+      for (Block* block : node->blocks()) {
+        handleBlockAndSubblocks(block);
+      }
+    }
+
+    // Processing for the block itself
+    std::unordered_map<Value*, std::vector<Node*>> grouped_linear_layers;
+    std::vector<Value*> ordered_tensor_inputs;
+    collectConstantLinearLayers(
+        block, grouped_linear_layers, ordered_tensor_inputs);
+
+    // Reverse topological ordering is used to prevent the need to
+    // update the aliasDB
+    for (auto tensor_it = ordered_tensor_inputs.rbegin();
+         tensor_it != ordered_tensor_inputs.rend();
+         ++tensor_it) {
+      collectAndMergeLinearLayers(grouped_linear_layers.at(*tensor_it));
+    }
+  }
+
+ public:
+  bool run(std::shared_ptr<Graph> graph) {
+    graph_ = graph;
```

Comment:
nit: the conventions with other passes is to pass `ConcatLinearLayers(Graph)` and then have `run` not take in a graph. 

Currently, you allow passing in instantiating the class, passing in one `Graph` via run, and then passing in a second `Graph` into the same instantiated class. `graph_` will get reset, but not the other member fields, so you'd using the wrong `AliasDb` and `graph_modified` would be stale.